llama-cpp.nix

```diff
@@ -0,0 +1,118 @@
+{ lib
+, stdenv
+, cmake
+, ninja
+, pkg-config
+, symlinkJoin
+, python3
+, openmpi
+, openblas
+, cudaPackages
+, rocmPackages
+, clblast
+, Accelerate ? null
+, MetalKit ? null
+, CoreVideo ? null
+, CoreGraphics ? null
+, useOpenCL ? false
+, useCuda ? false
+, useRocm ? false
+}:
+
+let
+  inherit (lib) optional optionals versionOlder;
+  isDefault = !useOpenCL && !useCuda && !useRocm;
+
+  # Give a little description difference between the flavors.
+  descriptionSuffix = if useOpenCL then
+    " (OpenCL accelerated)"
+  else if useCuda then
+    " (CUDA accelerated)"
+  else if useRocm then
+    " (ROCm accelerated)"
+  else if (MetalKit != null) then
+    " (MetalKit accelerated)"
+  else "";
+
+  llama-python = python3.withPackages (ps: [ ps.numpy ps.sentencepiece ]);
+
+  # See ./overlay.nix for where these dependencies are passed in.
+  defaultBuildInputs = builtins.filter (p: p != null) [
+    Accelerate
+    MetalKit
+    CoreVideo
+    CoreGraphics
+  ];
+
+  cudaBuildInput = symlinkJoin (with cudaPackages; {
+    # HACK(Green-Sky): Nix currently has issues with CMake findcudatoolkit
+    # see https://github.com/NixOS/nixpkgs/issues/224291
+    # copied from jaxlib
+    name = "${cudatoolkit.name}-merged";
+    paths = [
+      cudatoolkit.lib
+      cudatoolkit.out
+    ] ++ optionals (versionOlder cudatoolkit.version "11") [
+      # For some reason, some of the required libs are in the targets/x86_64-linux
+      # directory; not sure why, but this works around it.
+      "${cudatoolkit}/targets/${stdenv.system}"
+    ];
+  });
+
+  rocmBuildInputs = with rocmPackages; [ clr hipblas rocblas ];
+in
+
+stdenv.mkDerivation {
```

Comment:
Maybe out of scope of this PR, but it'd be nice to migrate to `buildPythonPackage`, as well as ensure this repo is pep517-compatible, and ensure the correct `.dist-info` is generated

Reply:
Ah, nevermind, I see now that the python bindings are in a separate project